iohk/NixOps.hs

```diff
@@ -858,6 +885,72 @@ date o c = parallelIO o c $
   \n -> ssh' o c "date" [] n
   (\out -> TIO.putStrLn $ fromNodeName n <> ": " <> out)
 
+
+s3Upload :: T.Text -> NixopsConfig -> IO ()
+s3Upload daedalus_rev c = do
+  let
+    say = liftIO . TIO.putStrLn
+    uploadOneFile :: BucketName -> T.Text -> ObjectKey -> Integer -> T.Text -> AWST (ResourceT IO) ()
+    uploadOneFile bucketName localPath remoteKey appver cardanoCommit = do
+      bdy <- chunkedFile defaultChunkSize (T.unpack localPath)
+      let
+        newMetaData :: HMS.HashMap Text Text
+        newMetaData = HMS.fromList [
+              ("daedalus-revision", daedalus_rev)
+            , ("cardano-revision", cardanoCommit)
+            , ("application-version", (T.pack. show) appver)
+          ]
+      void . send $ Lens.set poACL (Just OPublicRead) $ Lens.set poMetadata newMetaData $ putObject bucketName remoteKey bdy
+    copyObject' :: BucketName -> T.Text -> ObjectKey -> AWST (ResourceT IO) ()
+    copyObject' bucketName source dest = void . send $ Lens.set coACL (Just OPublicRead) $ copyObject bucketName source dest
+    uploadHashedInstaller :: T.Text -> T.Text -> Integer -> T.Text -> AWST (ResourceT IO) T.Text
+    uploadHashedInstaller bucketName localPath appver cardanoCommit = do
+      hash <- (liftIO . hashInstaller) localPath
+      let
+        basename' = last $ T.splitOn "/" $ localPath
```

Comment:
`last` is partial, and as such will blow up if the list is empty, but luckily for us `splitOn` always returns at least one result. Maybe just add a quick comment explaining we are safe here?